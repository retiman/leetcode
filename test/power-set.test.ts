describe('power set', () => {
  function __ps(xs: Set<number>): Set<Set<number>> {
    if (xs.size === 0) {
      return new Set(new Set());
    }

    // Find the first element of xs and call it x; remove it from the list.
    const x = xs.values().next().value;

    // Generate all subsets without x.
    const rest = new Set(xs);
    rest.delete(x);

    const excluded: Set<Set<number>> = __ps(rest);

    // Add x to all of the subsets from above.
    const included: Set<Set<number>> = new Set();
    excluded.forEach(ex => {
      const rest = new Set(ex);
      rest.add(x);

      included.add(rest);
    });

    // The subsets with x and the subsets without x constitute the power set.
    return new Set([...excluded, ...included]);
  }

  function ps(xs: Array<number>) {
    return __ps(new Set(xs));
  }

  test('run', async () => {
    expect(ps([])).toMatchSnapshot();
    expect(ps([1])).toMatchSnapshot();
  });
});
