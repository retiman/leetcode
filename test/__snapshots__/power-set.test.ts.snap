// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`power set run 1`] = `
Set {
  Set {},
}
`;

exports[`power set run 2`] = `
Set {
  Set {},
  Set {
    1,
  },
}
`;

exports[`power set run 3`] = `
Set {
  Set {},
  Set {
    2,
  },
  Set {
    1,
  },
  Set {
    2,
    1,
  },
}
`;

exports[`power set run 4`] = `
Set {
  Set {},
  Set {
    3,
  },
  Set {
    2,
  },
  Set {
    3,
    2,
  },
  Set {
    1,
  },
  Set {
    3,
    1,
  },
  Set {
    2,
    1,
  },
  Set {
    3,
    2,
    1,
  },
}
`;
